File: practice_anthropic/practice/q19_mapreduce.cpp
/*
 * Q19 — MapReduce Framework [Medium-Hard]
 *
 * Topic: Parallel Programming — MapReduce
 *
 * Problem:
 *   Implement a simplified single-machine MapReduce framework that can
 *   process data through three phases: Map, Shuffle, and Reduce.
 *
 *   Implement:
 *   1. map_phase(input_data, mapper_fn) — applies mapper to each input,
 *      produces a flat list of (key, value) pairs
 *   2. shuffle_phase(mapped_data) — groups all values by key
 *   3. reduce_phase(shuffled_data, reducer_fn) — applies reducer to each
 *      key's value list, producing a single (key, result) per key
 *   4. mapreduce(input_data, mapper_fn, reducer_fn) — orchestrates all phases
 *
 *   Then use the framework to implement:
 *   - Word count: count occurrences of each word in a list of strings
 *   - Inverted index: for each word, list which document IDs contain it
 *   - Mean computation: compute average of values grouped by key
 *
 * Constraints:
 *   - Mapper takes one input and returns a vector of (key, value) pairs
 *   - Reducer takes a key and a vector of values and returns a single value
 *   - Shuffle must be stable (preserve insertion order of values per key)
 *   - Handle empty inputs gracefully
 *
 * Key Concepts:
 *   - MapReduce paradigm (Google, 2004)
 *   - Functional decomposition of data processing
 *   - Key-value aggregation patterns
 *   - Foundation for distributed computing frameworks (Hadoop, Spark)
 */

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <vector>
#include <string>
#include <map>
#include <functional>
#include <sstream>
#include <algorithm>
#include <cmath>
#include <set>

using namespace std;

// A key-value pair used throughout the MapReduce pipeline
struct KVPair {
    string key;
    string value;
};

// ============ YOUR IMPLEMENTATION ============

/*
 * map_phase: Apply mapper_fn to each element of input_data.
 * mapper_fn takes a single string input and returns a vector of KVPairs.
 * Return the flattened list of all KVPairs produced.
 */
vector<KVPair> map_phase(const vector<string>& input_data,
                         function<vector<KVPair>(const string&)> mapper_fn) {
    vector<KVPair> result;

    // TODO: For each input in input_data:
    //   1. Call mapper_fn(input)
    //   2. Append all returned KVPairs to result

    return result;
}

/*
 * shuffle_phase: Group all values by key.
 * Returns a map from key -> list of values (in the order they appeared).
 */
map<string, vector<string>> shuffle_phase(const vector<KVPair>& mapped_data) {
    map<string, vector<string>> grouped;

    // TODO: For each KVPair in mapped_data:
    //   Group values under their key, preserving order

    return grouped;
}

/*
 * reduce_phase: Apply reducer_fn to each key's list of values.
 * reducer_fn takes (key, values) and returns a single string result.
 * Returns a map from key -> reduced result.
 */
map<string, string> reduce_phase(const map<string, vector<string>>& shuffled_data,
                                 function<string(const string&, const vector<string>&)> reducer_fn) {
    map<string, string> result;

    // TODO: For each (key, values) in shuffled_data:
    //   result[key] = reducer_fn(key, values)

    return result;
}

/*
 * mapreduce: Orchestrate all three phases.
 */
map<string, string> mapreduce(const vector<string>& input_data,
                              function<vector<KVPair>(const string&)> mapper_fn,
                              function<string(const string&, const vector<string>&)> reducer_fn) {
    // TODO: Chain map_phase -> shuffle_phase -> reduce_phase
    map<string, string> result;
    return result;
}

// --- Application-specific mappers and reducers ---

/*
 * word_count_mapper: Split input string into words, emit (word, "1") for each.
 */
vector<KVPair> word_count_mapper(const string& input) {
    vector<KVPair> pairs;

    // TODO: Split input by whitespace
    //   For each word, emit KVPair{word, "1"}

    return pairs;
}

/*
 * word_count_reducer: Sum up all the "1"s for a given word.
 * Return the count as a string.
 */
string word_count_reducer(const string& key, const vector<string>& values) {
    // TODO: Return to_string(values.size()) or sum of integer values
    return "0";
}

/*
 * inverted_index_mapper: Input is "doc_id:text". Emit (word, doc_id) for
 * each unique word in the text.
 */
vector<KVPair> inverted_index_mapper(const string& input) {
    vector<KVPair> pairs;

    // TODO: Parse "doc_id:text" format
    //   Split text into words
    //   For each unique word, emit KVPair{word, doc_id}

    return pairs;
}

/*
 * inverted_index_reducer: Concatenate doc IDs into a comma-separated list.
 */
string inverted_index_reducer(const string& key, const vector<string>& values) {
    // TODO: Join values with ","
    return "";
}

/*
 * mean_mapper: Input is "group:value". Emit (group, value).
 */
vector<KVPair> mean_mapper(const string& input) {
    vector<KVPair> pairs;

    // TODO: Parse "group:value" and emit KVPair{group, value}

    return pairs;
}

/*
 * mean_reducer: Compute the arithmetic mean of all values for a key.
 * Return the mean as a string with 2 decimal places.
 */
string mean_reducer(const string& key, const vector<string>& values) {
    // TODO: Parse each value as a double, compute mean
    //   Return formatted string (e.g., "3.50")
    return "0.00";
}

// ============ TEST FRAMEWORK ============

int main() {
    printf("=== Q19: MapReduce Framework ===\n\n");
    int passed = 0, total = 0;

    // Test 1: Word count on sample text
    {
        total++;
        vector<string> docs = {
            "the cat sat on the mat",
            "the dog sat on the log",
            "the cat and the dog"
        };
        auto result = mapreduce(docs, word_count_mapper, word_count_reducer);

        bool ok = (result["the"] == "5") &&
                  (result["cat"] == "2") &&
                  (result["sat"] == "2") &&
                  (result["dog"] == "2") &&
                  (result["on"] == "2") &&
                  (result["mat"] == "1") &&
                  (result["log"] == "1") &&
                  (result["and"] == "1");

        printf("%s [Word count: 'the'=%s(exp 5), 'cat'=%s(exp 2), 'mat'=%s(exp 1)]\n",
               ok ? "PASS" : "FAIL",
               result["the"].c_str(), result["cat"].c_str(), result["mat"].c_str());
        if (ok) passed++;
    }

    // Test 2: Inverted index
    {
        total++;
        vector<string> docs = {
            "0:hello world",
            "1:hello there",
            "2:world peace"
        };
        auto result = mapreduce(docs, inverted_index_mapper, inverted_index_reducer);

        bool ok = (result["hello"] == "0,1") &&
                  (result["world"] == "0,2") &&
                  (result["there"] == "1") &&
                  (result["peace"] == "2");

        printf("%s [Inverted index: 'hello'=%s(exp 0,1), 'world'=%s(exp 0,2)]\n",
               ok ? "PASS" : "FAIL",
               result["hello"].c_str(), result["world"].c_str());
        if (ok) passed++;
    }

    // Test 3: Mean computation
    {
        total++;
        vector<string> data = {
            "A:10", "A:20", "A:30",
            "B:5", "B:15",
            "C:100"
        };
        auto result = mapreduce(data, mean_mapper, mean_reducer);

        bool ok = (result["A"] == "20.00") &&
                  (result["B"] == "10.00") &&
                  (result["C"] == "100.00");

        printf("%s [Mean: A=%s(exp 20.00), B=%s(exp 10.00), C=%s(exp 100.00)]\n",
               ok ? "PASS" : "FAIL",
               result["A"].c_str(), result["B"].c_str(), result["C"].c_str());
        if (ok) passed++;
    }

    // Test 4: Empty input handling
    {
        total++;
        vector<string> empty_data;
        auto result = mapreduce(empty_data, word_count_mapper, word_count_reducer);
        bool ok = result.empty();

        printf("%s [Empty input returns empty result (size=%zu)]\n",
               ok ? "PASS" : "FAIL", result.size());
        if (ok) passed++;
    }

    // Test 5: Large-ish input (100+ items)
    {
        total++;
        vector<string> data;
        for (int i = 0; i < 120; i++) {
            data.push_back("apple banana");
        }
        data.push_back("cherry");

        auto result = mapreduce(data, word_count_mapper, word_count_reducer);

        bool ok = (result["apple"] == "120") &&
                  (result["banana"] == "120") &&
                  (result["cherry"] == "1");

        printf("%s [Large input (121 items): apple=%s(exp 120), cherry=%s(exp 1)]\n",
               ok ? "PASS" : "FAIL",
               result["apple"].c_str(), result["cherry"].c_str());
        if (ok) passed++;
    }

    // Test 6: Custom mapper/reducer — character frequency
    {
        total++;
        auto char_mapper = [](const string& input) -> vector<KVPair> {
            vector<KVPair> pairs;
            for (char c : input) {
                if (c != ' ') {
                    pairs.push_back({string(1, c), "1"});
                }
            }
            return pairs;
        };
        auto sum_reducer = [](const string& key, const vector<string>& values) -> string {
            return to_string(values.size());
        };

        vector<string> data = {"aab", "bcc"};
        auto result = mapreduce(data, char_mapper, sum_reducer);

        bool ok = (result["a"] == "2") &&
                  (result["b"] == "2") &&
                  (result["c"] == "2");

        printf("%s [Custom char-frequency: a=%s(exp 2), b=%s(exp 2), c=%s(exp 2)]\n",
               ok ? "PASS" : "FAIL",
               result["a"].c_str(), result["b"].c_str(), result["c"].c_str());
        if (ok) passed++;
    }

    printf("\nPassed %d/%d tests\n", passed, total);
    return (passed == total) ? 0 : 1;
}

File: practice_anthropic/practice/q19_mapreduce.py
"""
Q19 — MapReduce Framework [Medium-Hard]

Topic: Parallel Programming — MapReduce

Problem:
  Implement a simplified single-machine MapReduce framework that can
  process data through three phases: Map, Shuffle, and Reduce.

  Implement:
  1. map_phase(input_data, mapper_fn) — applies mapper to each input,
     produces a flat list of (key, value) pairs
  2. shuffle_phase(mapped_data) — groups all values by key
  3. reduce_phase(shuffled_data, reducer_fn) — applies reducer to each
     key's value list, producing a single (key, result) per key
  4. mapreduce(input_data, mapper_fn, reducer_fn) — orchestrates all phases

  Then use the framework to implement:
  - Word count: count occurrences of each word in a list of strings
  - Inverted index: for each word, list which document IDs contain it
  - Mean computation: compute average of values grouped by key

Constraints:
  - Mapper takes one input and returns a list of (key, value) tuples
  - Reducer takes a key and a list of values and returns a single value
  - Shuffle must be stable (preserve insertion order of values per key)
  - Handle empty inputs gracefully

Key Concepts:
  - MapReduce paradigm (Google, 2004)
  - Functional decomposition of data processing
  - Key-value aggregation patterns
  - Foundation for distributed computing frameworks (Hadoop, Spark)
"""

from collections import OrderedDict


# ============ YOUR IMPLEMENTATION ============

def map_phase(input_data, mapper_fn):
    """
    Apply mapper_fn to each element of input_data.
    mapper_fn takes a single string and returns a list of (key, value) tuples.
    Return the flattened list of all (key, value) tuples produced.
    """
    ret = []
    # TODO: For each item in input_data:
    for item in input_data:
        #   1. Call mapper_fn(item)
        #   2. Extend the result list with all returned tuples
        ret.extend(mapper_fn(item))
    return ret


def shuffle_phase(mapped_data):
    """
    Group all values by key, preserving insertion order.
    Returns an OrderedDict from key -> list of values.
    """
    ord_dict = OrderedDict()
    # TODO: For each (key, value) in mapped_data:
    for key, val in mapped_data:
        #   Group values under their key, preserving order
        if key not in ord_dict:
            ord_dict[key] = []
        ord_dict[key].append(val)

    return ord_dict


def reduce_phase(shuffled_data, reducer_fn):
    """
    Apply reducer_fn to each key's list of values.
    reducer_fn takes (key, values) and returns a single result.
    Returns a dict from key -> reduced result.
    """
    result = {}

    # TODO: For each (key, values) in shuffled_data:
    for key, vals in shuffled_data.items():
        result[key] = reducer_fn(key, vals)

    return result


def mapreduce(input_data, mapper_fn, reducer_fn):
    """Orchestrate all three phases: map -> shuffle -> reduce."""
    # TODO: Chain map_phase -> shuffle_phase -> reduce_phase
    mapped_data = map_phase(input_data, mapper_fn)
    shuffled_data = shuffle_phase(mapped_data)
    return reduce_phase(shuffled_data, reducer_fn)


# --- Application-specific mappers and reducers ---

def word_count_mapper(text):
    """Split text into words, emit (word, 1) for each."""
    # TODO: return [(word, 1) for word in text.split()]
    return [(word, 1) for word in text.split()]


def word_count_reducer(key, values):
    """Sum up all the 1s for a given word."""
    # TODO: return sum(values)
    return sum(values)


def inverted_index_mapper(doc_input):
    """
    Input is "doc_id:text". Emit (word, doc_id) for each unique word.
    """
    # TODO: Parse "doc_id:text", split text into words
    #   For each unique word, emit (word, doc_id)
    # partition will result in [doc_id, ":", text]
    doc_id, _, text = doc_input.partition(":")
    return [(word, doc_id) for word in text.split()]


def inverted_index_reducer(key, values):
    """Return list of doc IDs (as-is)."""
    # TODO: return values (already a list of doc_ids)
    return values


def mean_mapper(entry):
    """Input is "group:value". Emit (group, float(value))."""
    # TODO: Parse "group:value" and emit [(group, float(value))]
    group, _, value = entry.partition(":")
    return [(group, float(value))]


def mean_reducer(key, values):
    """Compute arithmetic mean of all values."""
    # TODO: return sum(values) / len(values)
    return sum(values) / len(values)


# ============ TEST FRAMEWORK ============

if __name__ == "__main__":
    print("=== Q19: MapReduce Framework ===\n")
    passed = 0
    total = 0

    # Test 1: Word count on sample text
    total += 1
    docs = [
        "the cat sat on the mat",
        "the dog sat on the log",
        "the cat and the dog",
    ]
    result = mapreduce(docs, word_count_mapper, word_count_reducer)
    ok = (result.get("the") == 6 and result.get("cat") == 2 and
          result.get("sat") == 2 and result.get("dog") == 2 and
          result.get("on") == 2 and result.get("mat") == 1 and
          result.get("log") == 1 and result.get("and") == 1)
    print(f"{'PASS' if ok else 'FAIL'} [Word count: the={result.get('the')}(exp 6), "
          f"cat={result.get('cat')}(exp 2), mat={result.get('mat')}(exp 1)]")
    if ok:
        passed += 1

    # Test 2: Inverted index
    total += 1
    docs = ["0:hello world", "1:hello there", "2:world peace"]
    result = mapreduce(docs, inverted_index_mapper, inverted_index_reducer)
    ok = (sorted(result.get("hello", [])) == ["0", "1"] and
          sorted(result.get("world", [])) == ["0", "2"] and
          result.get("there") == ["1"] and
          result.get("peace") == ["2"])
    print(f"{'PASS' if ok else 'FAIL'} [Inverted index: hello={result.get('hello')}(exp ['0','1']), "
          f"world={result.get('world')}(exp ['0','2'])]")
    if ok:
        passed += 1

    # Test 3: Mean computation
    total += 1
    data = ["A:10", "A:20", "A:30", "B:5", "B:15", "C:100"]
    result = mapreduce(data, mean_mapper, mean_reducer)
    ok = (abs(result.get("A", 0) - 20.0) < 0.01 and
          abs(result.get("B", 0) - 10.0) < 0.01 and
          abs(result.get("C", 0) - 100.0) < 0.01)
    print(f"{'PASS' if ok else 'FAIL'} [Mean: A={result.get('A')}(exp 20.0), "
          f"B={result.get('B')}(exp 10.0), C={result.get('C')}(exp 100.0)]")
    if ok:
        passed += 1

    # Test 4: Empty input handling
    total += 1
    result = mapreduce([], word_count_mapper, word_count_reducer)
    ok = len(result) == 0
    print(f"{'PASS' if ok else 'FAIL'} [Empty input returns empty result (size={len(result)})]")
    if ok:
        passed += 1

    # Test 5: Large-ish input (100+ items)
    total += 1
    data = ["apple banana"] * 120 + ["cherry"]
    result = mapreduce(data, word_count_mapper, word_count_reducer)
    ok = (result.get("apple") == 120 and result.get("banana") == 120 and
          result.get("cherry") == 1)
    print(f"{'PASS' if ok else 'FAIL'} [Large input (121 items): "
          f"apple={result.get('apple')}(exp 120), cherry={result.get('cherry')}(exp 1)]")
    if ok:
        passed += 1

    # Test 6: Custom mapper/reducer — character frequency
    total += 1
    char_mapper = lambda s: [(c, 1) for c in s if c != " "]
    sum_reducer = lambda key, vals: sum(vals)
    data = ["aab", "bcc"]
    result = mapreduce(data, char_mapper, sum_reducer)
    ok = (result.get("a") == 2 and result.get("b") == 2 and result.get("c") == 2)
    print(f"{'PASS' if ok else 'FAIL'} [Custom char-frequency: "
          f"a={result.get('a')}(exp 2), b={result.get('b')}(exp 2), c={result.get('c')}(exp 2)]")
    if ok:
        passed += 1

    print(f"\nPassed {passed}/{total} tests")
